File: simple_pipe/builder/pipeline_builder.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable

from simple_pipe.enums import DropPolicy, NodeRole
from simple_pipe.operators.base import Operator
from simple_pipe.operators.generic.app_des import AppSink
from simple_pipe.operators.generic.app_src import AppSource
from simple_pipe.operators.registry import NodeFactory, default_node_factory
from simple_pipe.result import Result
from simple_pipe.spec.graph_spec import PipelineGraphSpec


@dataclass
class FlowChartNodeSpec:
    id: str
    type: str
    init_params: dict = field(default_factory=dict)
    upstream_ids: list[str] = field(default_factory=list)


@dataclass
class FlowChartPipelineSpec:
    nodes: list[FlowChartNodeSpec]

    def by_id(self, node_id: str) -> FlowChartNodeSpec | None:
        for node in self.nodes:
            if node.id == node_id:
                return node
        return None


@dataclass
class BuiltPipeline:
    source: AppSource
    sink: AppSink
    nodes: list[Operator]
    node_map: dict[str, Operator]


TopologyValidator = Callable[[FlowChartPipelineSpec], Result[None]]


def _shallow_merge(base: dict, overlay: dict) -> dict:
    merged = dict(base)
    merged.update(overlay)
    return merged


def load_node_defaults(defaults_dir: str | Path, node_type: str) -> dict:
    path = Path(defaults_dir) / f"{node_type}.json"
    if not path.exists():
        return {}
    return json.loads(path.read_text(encoding="utf-8"))


def build_flow_chart_pipeline_spec_from_graph(
    graph: PipelineGraphSpec,
    load_defaults: Callable[[str], dict] | None = None,
) -> FlowChartPipelineSpec:
    load_defaults = load_defaults or (lambda _t: {})
    upstream_map: dict[str, list[str]] = {n.id: [] for n in graph.nodes}
    for edge in graph.edges:
        if edge.to not in upstream_map:
            upstream_map[edge.to] = []
        upstream_map[edge.to].extend(edge.from_nodes)

    flow_nodes: list[FlowChartNodeSpec] = []
    for node in graph.nodes:
        defaults = load_defaults(node.type)
        init_params = _shallow_merge(defaults, node.init_params)
        flow_nodes.append(
            FlowChartNodeSpec(
                id=node.id,
                type=node.type,
                init_params=init_params,
                upstream_ids=upstream_map.get(node.id, []),
            )
        )
    return FlowChartPipelineSpec(nodes=flow_nodes)


def validate_flow_chart_pipeline_spec(
    spec: FlowChartPipelineSpec,
    domain_rules: TopologyValidator | None = None,
) -> Result[None]:
    ids = {n.id for n in spec.nodes}
    if len(ids) != len(spec.nodes):
        return Result.failure("Duplicate node id in graph")

    for node in spec.nodes:
        for upstream in node.upstream_ids:
            if upstream not in ids:
                return Result.failure(f"Edge references unknown upstream: {upstream}")

    roles = {}
    for instance in spec.nodes:
        # role determined after instantiation; validate structural rules on types
        if instance.type == "app_src" and instance.upstream_ids:
            return Result.failure("app_src must not have upstream edges")
        if instance.type == "app_des":
            pass  # may have upstreams only

    sources = [n for n in spec.nodes if n.type == "app_src"]
    sinks = [n for n in spec.nodes if n.type == "app_des"]
    if len(sources) != 1:
        return Result.failure("Graph must contain exactly one app_src")
    if len(sinks) != 1:
        return Result.failure("Graph must contain exactly one app_des")

    if domain_rules:
        return domain_rules(spec)
    return Result.success()


def build_flow_chart_pipeline_from_spec(
    spec: FlowChartPipelineSpec,
    node_factory: NodeFactory | None = None,
) -> BuiltPipeline:
    factory = node_factory or default_node_factory()
    node_map: dict[str, Operator] = {}

    for node_spec in spec.nodes:
        op = factory(node_spec.id, node_spec.type, node_spec.init_params)
        node_map[node_spec.id] = op

    for node_spec in spec.nodes:
        op = node_map[node_spec.id]
        upstreams = [node_map[uid] for uid in node_spec.upstream_ids]
        op.attach_to(upstreams)
        for upstream in upstreams:
            upstream.add_downstream(op)

    sources = [n for n in node_map.values() if n.node_role() == NodeRole.SRC]
    sinks = [n for n in node_map.values() if n.node_role() == NodeRole.DES]
    if len(sources) != 1 or len(sinks) != 1:
        raise RuntimeError("Built graph must have exactly one source and one sink")
    source = sources[0]
    sink = sinks[0]
    if not isinstance(source, AppSource) or not isinstance(sink, AppSink):
        raise RuntimeError("Source/sink type mismatch")

    return BuiltPipeline(
        source=source,
        sink=sink,
        nodes=list(node_map.values()),
        node_map=node_map,
    )


def apply_pipeline_queue_limits(
    built: BuiltPipeline,
    pipeline_params: dict,
    drop_policy: DropPolicy = DropPolicy.DROP_NEWEST,
) -> None:
    src_limit = int(
        pipeline_params.get("app_src_in_queue_limit")
        or pipeline_params.get("source_queue_limit")
        or 8
    )
    des_limit = int(
        pipeline_params.get("app_des_in_queue_limit")
        or pipeline_params.get("sink_queue_limit")
        or 8
    )
    built.source.set_input_queue_limit(src_limit, drop_policy)
    built.sink.set_input_queue_limit(des_limit, drop_policy)

File: src/builder/pipeline_builder.cpp
#include "simple_pipe/builder/pipeline_builder.hpp"

#include <fstream>
#include <unordered_set>

namespace simple_pipe {

namespace {

bool IsIngressType(const std::string& type) {
  return type == "app_src" || type == "image_src" || type == "video_src";
}

nlohmann::json ShallowMerge(const nlohmann::json& base, const nlohmann::json& overlay) {
  nlohmann::json merged = base.is_object() ? base : nlohmann::json::object();
  if (overlay.is_object()) {
    for (auto it = overlay.begin(); it != overlay.end(); ++it) {
      merged[it.key()] = it.value();
    }
  }
  return merged;
}

}  // namespace

const FlowChartNodeSpec* FlowChartPipelineSpec::Find(const std::string& id) const {
  for (const auto& node : nodes) {
    if (node.id == id) {
      return &node;
    }
  }
  return nullptr;
}

nlohmann::json LoadNodeDefaults(const std::string& defaults_dir, const std::string& node_type) {
  const std::string path = defaults_dir + "/" + node_type + ".json";
  std::ifstream in(path);
  if (!in) {
    return nlohmann::json::object();
  }
  nlohmann::json data;
  in >> data;
  return data;
}

FlowChartPipelineSpec BuildFlowChartPipelineSpecFromGraph(const PipelineGraphSpec& graph,
                                                           DefaultsLoader load_defaults) {
  std::unordered_map<std::string, std::vector<std::string>> upstream_map;
  for (const auto& node : graph.nodes) {
    upstream_map[node.id] = {};
  }
  for (const auto& edge : graph.edges) {
    upstream_map[edge.to].insert(upstream_map[edge.to].end(), edge.from_nodes.begin(),
                                 edge.from_nodes.end());
  }

  FlowChartPipelineSpec spec;
  for (const auto& node : graph.nodes) {
    nlohmann::json defaults = nlohmann::json::object();
    if (load_defaults) {
      defaults = load_defaults(node.type);
    }
    FlowChartNodeSpec flow;
    flow.id = node.id;
    flow.type = node.type;
    flow.init_params = ShallowMerge(defaults, node.init_params);
    flow.upstream_ids = upstream_map[node.id];
    spec.nodes.push_back(std::move(flow));
  }
  return spec;
}

Result<void> ValidateFlowChartPipelineSpec(const FlowChartPipelineSpec& spec,
                                           TopologyValidator domain_rules) {
  std::unordered_set<std::string> ids;
  for (const auto& node : spec.nodes) {
    if (!ids.insert(node.id).second) {
      return Result<void>::failure("Duplicate node id in graph");
    }
  }

  for (const auto& node : spec.nodes) {
    for (const auto& upstream : node.upstream_ids) {
      if (ids.find(upstream) == ids.end()) {
        return Result<void>::failure("Edge references unknown upstream: " + upstream);
      }
    }
    if (IsIngressType(node.type) && !node.upstream_ids.empty()) {
      return Result<void>::failure("ingress source must not have upstream edges: " + node.type);
    }
  }

  int source_count = 0;
  int sink_count = 0;
  for (const auto& node : spec.nodes) {
    if (IsIngressType(node.type)) {
      ++source_count;
    }
    if (node.type == "app_des") {
      ++sink_count;
    }
  }
  if (source_count != 1) {
    return Result<void>::failure(
        "Graph must contain exactly one ingress source (app_src | image_src | video_src)");
  }
  if (sink_count != 1) {
    return Result<void>::failure("Graph must contain exactly one app_des");
  }

  if (domain_rules) {
    return domain_rules(spec);
  }
  return Result<void>::success();
}

BuiltPipeline BuildFlowChartPipelineFromSpec(const FlowChartPipelineSpec& spec, NodeFactory factory) {
  if (!factory) {
    factory = DefaultNodeFactory();
  }

  BuiltPipeline built;
  for (const auto& node_spec : spec.nodes) {
    auto op = factory(node_spec.id, node_spec.type, node_spec.init_params);
    built.node_map[node_spec.id] = op;
    built.nodes.push_back(op);
  }

  for (const auto& node_spec : spec.nodes) {
    auto op = built.node_map.at(node_spec.id);
    std::vector<std::shared_ptr<Operator>> upstreams;
    for (const auto& uid : node_spec.upstream_ids) {
      upstreams.push_back(built.node_map.at(uid));
    }
    op->AttachTo(upstreams);
    for (const auto& up : upstreams) {
      up->AddDownstream(op);
    }
  }

  for (const auto& node : built.nodes) {
    if (node->role() == NodeRole::kSrc) {
      built.source = std::dynamic_pointer_cast<SourceOperator>(node);
      if (built.source) {
        built.pull_driven = built.source->PullDriven();
      }
    }
    if (node->role() == NodeRole::kDes) {
      built.sink = std::dynamic_pointer_cast<AppSink>(node);
    }
  }

  if (!built.source || !built.sink) {
    throw std::runtime_error("Built graph must have exactly one ingress source and one AppSink");
  }
  return built;
}

void ApplyPipelineQueueLimits(BuiltPipeline& built, const nlohmann::json& pipeline_params,
                              DropPolicy policy) {
  int src_limit = 8;
  int des_limit = 8;
  if (pipeline_params.contains("app_src_in_queue_limit")) {
    src_limit = pipeline_params["app_src_in_queue_limit"].get<int>();
  } else if (pipeline_params.contains("source_queue_limit")) {
    src_limit = pipeline_params["source_queue_limit"].get<int>();
  }
  if (pipeline_params.contains("app_des_in_queue_limit")) {
    des_limit = pipeline_params["app_des_in_queue_limit"].get<int>();
  } else if (pipeline_params.contains("sink_queue_limit")) {
    des_limit = pipeline_params["sink_queue_limit"].get<int>();
  }
  built.source->SetInputQueueLimit(src_limit, policy);
  built.sink->SetInputQueueLimit(des_limit, policy);
}

}  // namespace simple_pipe
